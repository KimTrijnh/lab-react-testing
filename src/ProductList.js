import React from "react";
// import Btn from './button'
export default class ProductList extends React.Component {
  
  render() {
    return (
      <ul>
      {this.props.products.map(product => <li className="product-item">Price: ${product.price} USD 
      <button onClick={() => this.props.onProductBuy(product.id)}>Add to cart</button></li>)}
      </ul>
    );
  }
}
