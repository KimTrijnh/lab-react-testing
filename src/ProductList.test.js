import React from 'react';
import {shallow} from 'enzyme';
import ProductList from './ProductList'
// import { jet } from 'enzyme'


const TEST_DATA = [
    {id: 134, name: 'computer', price: 1000},
    {id: 122, name: 'anker mouse', price: 50.25},
    {id: 553, name: 'computer stand', price: 20.99},
    {id: 689, name: 'mac adapter', price: 10000}
  ];

  it('has 4 products', () => {
    const output = shallow(<
        ProductList products={TEST_DATA}/>)
    //test total product = 4
        expect(output.find('li').length).toEqual(4);
  })

  it('1st product contain 1000', () => {
    const output = shallow(<
        ProductList products={TEST_DATA}/>)
        const firstProduct = output.find('li').first()
    expect(firstProduct.text()).toMatch(/1000/)
  })

  it('item is rendered in form $XXX.00 USD', () => {
    const output = shallow(<
        ProductList products={TEST_DATA}/>)

    output.find('li').forEach(product => 
        expect(product.text()).toMatch(/\$\d+\.?\d?\d?\sUSD/))
  })
    



//if price is .00, then do not show the decimals. E.g. it should be $1000 USD or $50.25 USD.
// => 2 digits after . should be different from 00

it('product has className : product-item', () => {
    const output = shallow(<
        ProductList products={TEST_DATA}/>)
    output.find('li').forEach(product => 
        expect(product.hasClass('product-item')).toEqual(true))
})

it('product has 1 button', () => {
    const output = shallow(<
        ProductList products={TEST_DATA}/>)
    const firstProduct = output.find('li').first()
    expect(firstProduct.find('button').length).toBe(1)
})

it('call 1 when btn is clicked', () =>{
    const testingFn = jest.fn()
    const output = shallow(
        <ProductList products={TEST_DATA} 
            onProductBuy={testingFn} />
    )
    const firstProduct = output.find('button').first();
    firstProduct.simulate('click');
    expect(testingFn.mock.calls.length).toEqual(1);
    expect(testingFn.mock.calls[0][0]).toEqual(134);

})

it('matches our snapshot', () => {
    const output = shallow(
        <ProductList products={TEST_DATA} />
    )
    expect(output).toMatchSnapshot();
  });
  