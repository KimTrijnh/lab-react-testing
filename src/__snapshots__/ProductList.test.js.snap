// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches our snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ProductList
    products={
      Array [
        Object {
          "id": 134,
          "name": "computer",
          "price": 1000,
        },
        Object {
          "id": 122,
          "name": "anker mouse",
          "price": 50.25,
        },
        Object {
          "id": 553,
          "name": "computer stand",
          "price": 20.99,
        },
        Object {
          "id": 689,
          "name": "mac adapter",
          "price": 10000,
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <li
          className="product-item"
        >
          Price: $
          1000
           USD
          <button
            onClick={[Function]}
          >
            Add to cart
          </button>
        </li>,
        <li
          className="product-item"
        >
          Price: $
          50.25
           USD
          <button
            onClick={[Function]}
          >
            Add to cart
          </button>
        </li>,
        <li
          className="product-item"
        >
          Price: $
          20.99
           USD
          <button
            onClick={[Function]}
          >
            Add to cart
          </button>
        </li>,
        <li
          className="product-item"
        >
          Price: $
          10000
           USD
          <button
            onClick={[Function]}
          >
            Add to cart
          </button>
        </li>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "134",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Price: $",
            1000,
            " USD",
            <button
              onClick={[Function]}
            >
              Add to cart
            </button>,
          ],
          "className": "product-item",
        },
        "ref": null,
        "rendered": Array [
          "Price: $",
          1000,
          " USD",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Add to cart",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Add to cart",
            "type": "button",
          },
        ],
        "type": "li",
      },
      Object {
        "instance": null,
        "key": "122",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Price: $",
            50.25,
            " USD",
            <button
              onClick={[Function]}
            >
              Add to cart
            </button>,
          ],
          "className": "product-item",
        },
        "ref": null,
        "rendered": Array [
          "Price: $",
          50.25,
          " USD",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Add to cart",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Add to cart",
            "type": "button",
          },
        ],
        "type": "li",
      },
      Object {
        "instance": null,
        "key": "553",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Price: $",
            20.99,
            " USD",
            <button
              onClick={[Function]}
            >
              Add to cart
            </button>,
          ],
          "className": "product-item",
        },
        "ref": null,
        "rendered": Array [
          "Price: $",
          20.99,
          " USD",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Add to cart",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Add to cart",
            "type": "button",
          },
        ],
        "type": "li",
      },
      Object {
        "instance": null,
        "key": "689",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Price: $",
            10000,
            " USD",
            <button
              onClick={[Function]}
            >
              Add to cart
            </button>,
          ],
          "className": "product-item",
        },
        "ref": null,
        "rendered": Array [
          "Price: $",
          10000,
          " USD",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Add to cart",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Add to cart",
            "type": "button",
          },
        ],
        "type": "li",
      },
    ],
    "type": "ul",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <li
            className="product-item"
          >
            Price: $
            1000
             USD
            <button
              onClick={[Function]}
            >
              Add to cart
            </button>
          </li>,
          <li
            className="product-item"
          >
            Price: $
            50.25
             USD
            <button
              onClick={[Function]}
            >
              Add to cart
            </button>
          </li>,
          <li
            className="product-item"
          >
            Price: $
            20.99
             USD
            <button
              onClick={[Function]}
            >
              Add to cart
            </button>
          </li>,
          <li
            className="product-item"
          >
            Price: $
            10000
             USD
            <button
              onClick={[Function]}
            >
              Add to cart
            </button>
          </li>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "134",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Price: $",
              1000,
              " USD",
              <button
                onClick={[Function]}
              >
                Add to cart
              </button>,
            ],
            "className": "product-item",
          },
          "ref": null,
          "rendered": Array [
            "Price: $",
            1000,
            " USD",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Add to cart",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Add to cart",
              "type": "button",
            },
          ],
          "type": "li",
        },
        Object {
          "instance": null,
          "key": "122",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Price: $",
              50.25,
              " USD",
              <button
                onClick={[Function]}
              >
                Add to cart
              </button>,
            ],
            "className": "product-item",
          },
          "ref": null,
          "rendered": Array [
            "Price: $",
            50.25,
            " USD",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Add to cart",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Add to cart",
              "type": "button",
            },
          ],
          "type": "li",
        },
        Object {
          "instance": null,
          "key": "553",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Price: $",
              20.99,
              " USD",
              <button
                onClick={[Function]}
              >
                Add to cart
              </button>,
            ],
            "className": "product-item",
          },
          "ref": null,
          "rendered": Array [
            "Price: $",
            20.99,
            " USD",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Add to cart",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Add to cart",
              "type": "button",
            },
          ],
          "type": "li",
        },
        Object {
          "instance": null,
          "key": "689",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Price: $",
              10000,
              " USD",
              <button
                onClick={[Function]}
              >
                Add to cart
              </button>,
            ],
            "className": "product-item",
          },
          "ref": null,
          "rendered": Array [
            "Price: $",
            10000,
            " USD",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Add to cart",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Add to cart",
              "type": "button",
            },
          ],
          "type": "li",
        },
      ],
      "type": "ul",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
